internal/agent/device/resource/resource.go

```diff
@@ -62,18 +55,90 @@ func (m *ResourceManager) Run(ctx context.Context) {
 	defer m.log.Infof("Resource manager stopped")
 
 	go m.diskMonitor.Run(ctx)
-	m.cpuMonitor.Run(ctx)
+	go m.cpuMonitor.Run(ctx)
+	m.memoryMonitor.Run(ctx)
+}
+
+func (m *ResourceManager) Update(monitor v1alpha1.ResourceMonitor) (bool, error) {
+	monitorType, err := monitor.Discriminator()
+	if err != nil {
+		return false, err
+	}
+
+	switch monitorType {
+	case CPUMonitorType:
+		return m.cpuMonitor.Update(monitor)
+	case DiskMonitorType:
+		return m.diskMonitor.Update(monitor)
+	case MemoryMonitorType:
+		return m.memoryMonitor.Update(monitor)
+	default:
+		return false, fmt.Errorf("unknown monitor type: %s", monitorType)
+	}
+}
+
+func (m *ResourceManager) Alerts() *Alerts {
+	return &Alerts{
+		DiskUsage: m.diskMonitor.Alerts(),
+		CPUUsage:  m.cpuMonitor.Alerts(),
+	}
+}
+
+type Alerts struct {
+	DiskUsage   []v1alpha1.ResourceAlertRule
+	CPUUsage    []v1alpha1.ResourceAlertRule
+	MemoryUsage []v1alpha1.ResourceAlertRule
+}
 
+type Alert struct {
+	v1alpha1.ResourceAlertRule
+	duration    time.Duration
+	firing      bool
+	firingSince time.Time
 }
 
-func (m *ResourceManager) Usage() *Usage {
-	return &Usage{
-		DiskUsage: m.diskMonitor.Usage(),
-		CPUUsage:  m.cpuMonitor.Usage(),
+func NewAlert(rule v1alpha1.ResourceAlertRule) (*Alert, error) {
+	duration, err := time.ParseDuration(rule.Duration)
+	if err != nil {
+		return nil, err
 	}
+	return &Alert{
+		ResourceAlertRule: rule,
+		duration:          duration,
+	}, nil
+}
+
+func (a *Alert) IsFiring() bool {
+	return a.firing
+}
+
+var AlertLevelMap = map[v1alpha1.ResourceAlertSeverityType]AlertSeverity{
+	v1alpha1.ResourceAlertSeverityTypeInfo:     {Status: v1alpha1.DeviceResourceStatusHealthy, Level: 0},
+	v1alpha1.ResourceAlertSeverityTypeWarning:  {Status: v1alpha1.DeviceResourceStatusWarning, Level: 1},
+	v1alpha1.ResourceAlertSeverityTypeCritical: {Status: v1alpha1.DeviceResourceStatusCritical, Level: 2},
 }
 
-type Usage struct {
-	DiskUsage *DiskUsage
-	CPUUsage  *CPUUsage
+type AlertSeverity struct {
+	Status v1alpha1.DeviceResourceStatusType
+	Level  int
+}
+
+// GetHighestSeverityResourceStatusFromAlerts returns the highest severity statusDeviceResourceStatusType from a list of alerts.
+func GetHighestSeverityResourceStatusFromAlerts(alerts []v1alpha1.ResourceAlertRule) (v1alpha1.DeviceResourceStatusType, error) {
+	if len(alerts) == 0 {
+		return v1alpha1.DeviceResourceStatusUnknown, nil
```

Comment:
@celdrake

Reply:
Looks good, in the UI we did something similar to extract the message of the alert with the highest severity.
https://github.com/flightctl/flightctl-ui/pull/22/files#diff-e99c39011bc712d2766b296021bd3a9fac077341b017cbd2568455b6c7b4df47R40